vm/odroid/odroid.go

```diff
@@ -0,0 +1,353 @@
+	if err := inst.waitForSsh(10); err == nil {
+		Logf(1, "odroid: ssh succeeded, shutting down now")
+		inst.ssh("shutdown now")
+		if !vm.SleepInterruptible(20 * time.Second) {
+			return fmt.Errorf("shutdown in progress")
+		}
+	} else {
+		Logf(1, "odroid: ssh failed")
+	}
+
+	// Hard reset by turning off and back on power on a hub port.
+	Logf(1, "odroid: hard reset, turning off power")
+	if err := switchPortPower(inst.cfg.OdroidHubBus, inst.cfg.OdroidHubDevice, inst.cfg.OdroidHubPort, false); err != nil {
+		return err
+	}
+	if !vm.SleepInterruptible(5 * time.Second) {
+		return fmt.Errorf("shutdown in progress")
+	}
+	if err := switchPortPower(inst.cfg.OdroidHubBus, inst.cfg.OdroidHubDevice, inst.cfg.OdroidHubPort, true); err != nil {
+		return err
+	}
+
+	// Now wait for boot.
+	Logf(1, "odroid: power back on, waiting for boot")
+	if err := inst.waitForSsh(150); err != nil {
+		return err
+	}
+
+	Logf(1, "odroid: boot succeeded")
+	return nil
+}
+
+func (inst *instance) waitForSsh(timeout int) error {
+	var err error
+	start := time.Now()
+	for {
+		if !vm.SleepInterruptible(time.Second) {
+			return fmt.Errorf("shutdown in progress")
+		}
+		if _, err = inst.ssh("pwd"); err == nil {
+			return nil
+		}
+		if time.Since(start).Seconds() > float64(timeout) {
+			break
+		}
+	}
+	return fmt.Errorf("instance is dead and unrepairable: %v", err)
+}
+
+func (inst *instance) Close() {
+	close(inst.closed)
+	os.RemoveAll(inst.cfg.Workdir)
+}
+
+func (inst *instance) Copy(hostSrc string) (string, error) {
+	basePath := "/data/"
+	vmDst := filepath.Join(basePath, filepath.Base(hostSrc))
+	args := append(inst.sshArgs("-P"), hostSrc, "root@"+inst.cfg.OdroidSlaveAddr+":"+vmDst)
+	cmd := exec.Command("scp", args...)
+	if inst.cfg.Debug {
+		Logf(0, "running command: scp %#v", args)
+		cmd.Stdout = os.Stdout
+		cmd.Stderr = os.Stdout
+	}
+	if err := cmd.Start(); err != nil {
+		return "", err
+	}
+	done := make(chan bool)
+	go func() {
+		select {
+		case <-time.After(3 * time.Minute):
+			cmd.Process.Kill()
+		case <-done:
+		}
+	}()
+	err := cmd.Wait()
+	close(done)
+	if err != nil {
+		return "", err
+	}
+	return vmDst, nil
+}
+
+func (inst *instance) Run(timeout time.Duration, stop <-chan bool, command string) (<-chan []byte, <-chan error, error) {
+	tty, err := openConsole(inst.console)
+	if err != nil {
+		return nil, nil, err
+	}
+	inst.merger.Add("console", tty)
+
+	rpipe, wpipe, err := vm.LongPipe()
+	if err != nil {
+		return nil, nil, err
+	}
+	inst.merger.Add("ssh", rpipe)
+
+	args := append(inst.sshArgs("-p"), "root@"+inst.cfg.OdroidSlaveAddr, "cd /data; "+command)
+	if inst.cfg.Debug {
+		Logf(0, "running command: ssh %#v", args)
+	}
+	cmd := exec.Command("ssh", args...)
+	cmd.Stdout = wpipe
+	cmd.Stderr = wpipe
+	if err := cmd.Start(); err != nil {
+		wpipe.Close()
+		return nil, nil, err
+	}
+	wpipe.Close()
+	errc := make(chan error, 1)
+	signal := func(err error) {
+		select {
+		case errc <- err:
+		default:
+		}
+	}
+
+	go func() {
```

Comment:
I think you are leaking some file handles and goroutines here. And maybe parts of output.
tty needs to be closed somewhere. Read on it won't ever fail. So you will have the file descriptor opened and goroutine reading from it infinitely. tty object needs to be closed.
merger object needs to be Wait'ed somewhere. There is no point in putting merger into inst, since it's used only for duration of Run.
Also you need to handle the case when the command exits successfully. This is required for crash reproduction.
Take a look at adb's Run method. It should handle all of that properly.
Also as a litmus test run manager, kill ssh several times, and watch for leaked fd's in /proc/self/fd.